File: legacy-python/app/security.py
"""Auth: hash mật khẩu (PBKDF2 stdlib) + JWT access token."""
import hashlib
import hmac
import secrets
import time
from typing import Optional

import jwt
from fastapi import Depends, HTTPException, Request

from .config import ACCESS_TOKEN_TTL, JWT_ALGORITHM, SECRET_KEY
from .db import connect

PBKDF2_ITERATIONS = 200_000


def hash_password(password: str) -> str:
    salt = secrets.token_hex(16)
    digest = hashlib.pbkdf2_hmac(
        "sha256", password.encode(), bytes.fromhex(salt), PBKDF2_ITERATIONS
    ).hex()
    return f"pbkdf2${PBKDF2_ITERATIONS}${salt}${digest}"


def verify_password(password: str, stored: str) -> bool:
    try:
        _, iters, salt, digest = stored.split("$")
        candidate = hashlib.pbkdf2_hmac(
            "sha256", password.encode(), bytes.fromhex(salt), int(iters)
        ).hex()
        return hmac.compare_digest(candidate, digest)
    except (ValueError, AttributeError):
        return False


def create_token(user_id: str, role: str, token_version: int = 0) -> str:
    now = int(time.time())
    payload = {"sub": user_id, "role": role, "tv": token_version,
               "iat": now, "exp": now + ACCESS_TOKEN_TTL}
    return jwt.encode(payload, SECRET_KEY, algorithm=JWT_ALGORITHM)


def decode_token(token: str) -> Optional[dict]:
    try:
        return jwt.decode(token, SECRET_KEY, algorithms=[JWT_ALGORITHM])
    except jwt.PyJWTError:
        return None


def _extract_token(request: Request) -> Optional[str]:
    auth = request.headers.get("Authorization", "")
    if auth.startswith("Bearer "):
        return auth[7:].strip()
    return None


def _load_user(user_id: str) -> Optional[dict]:
    conn = connect()
    try:
        row = conn.execute("SELECT * FROM users WHERE id = ?", (user_id,)).fetchone()
        return dict(row) if row else None
    finally:
        conn.close()


def _token_fresh(payload: dict, user: dict) -> bool:
    """Token còn hiệu lực nếu tv trong token khớp token_version hiện tại của user
    (đổi mật khẩu/reset sẽ tăng token_version → mọi token cũ vô hiệu)."""
    return int(payload.get("tv", 0)) == int(user.get("token_version") or 0)


def get_current_user(request: Request) -> dict:
    token = _extract_token(request)
    payload = decode_token(token) if token else None
    # token trung gian (typ='2fa' chờ nhập mã) KHÔNG phải access token —
    # nếu không chặn ở đây thì 2FA bị vượt qua chỉ với mật khẩu
    if not payload or payload.get("typ"):
        raise HTTPException(401, "Chưa đăng nhập hoặc phiên đã hết hạn")
    user = _load_user(payload["sub"])
    if not user:
        raise HTTPException(401, "Tài khoản không tồn tại")
    if not _token_fresh(payload, user):
        raise HTTPException(401, "Phiên đã hết hạn — đăng nhập lại")
    return user


def get_optional_user(request: Request) -> Optional[dict]:
    token = _extract_token(request)
    payload = decode_token(token) if token else None
    if not payload or payload.get("typ"):
        return None
    user = _load_user(payload["sub"])
    if user and not _token_fresh(payload, user):
        return None
    return user


def create_pending_2fa_token(user_id: str) -> str:
    """Token trung gian sau khi đúng mật khẩu, chờ mã TOTP — sống 5 phút."""
    now = int(time.time())
    payload = {"sub": user_id, "typ": "2fa", "iat": now, "exp": now + 300}
    return jwt.encode(payload, SECRET_KEY, algorithm=JWT_ALGORITHM)


def require_admin(user: dict = Depends(get_current_user)) -> dict:
    if user.get("role") != "admin":
        raise HTTPException(403, "Cần quyền quản trị")
    return user


# ---- Phân quyền v2.0 ----
# admin    : toàn quyền (kể cả người dùng, cài đặt, đối tác)
# manager  : quản lý delivery — duyệt phát hành, chỉnh sửa mọi nội dung
# uploader : chỉ tạo/sửa product NHẠC CỦA MÌNH, không phát hành (gửi hàng đợi duyệt)
# user     : người nghe
STAFF_ROLES = ("admin", "manager", "uploader")
MANAGER_ROLES = ("admin", "manager")


def require_staff(user: dict = Depends(get_current_user)) -> dict:
    """Vào được khu vực CMS (admin/manager/uploader)."""
    if user.get("role") not in STAFF_ROLES:
        raise HTTPException(403, "Cần quyền quản trị nội dung")
    return user


def require_manager(user: dict = Depends(get_current_user)) -> dict:
    """Duyệt/phát hành/chỉnh sửa mọi nội dung (admin/manager)."""
    if user.get("role") not in MANAGER_ROLES:
        raise HTTPException(403, "Cần quyền quản lý phát hành")
    return user


def is_manager(user: dict) -> bool:
    return user.get("role") in MANAGER_ROLES

File: backend/src/security.ts
/** Auth — PBKDF2 (KHỚP ĐỊNH DẠNG bản Python để user cũ đăng nhập được) + JWT + RBAC. */
import crypto from 'node:crypto';
import jwt from 'jsonwebtoken';
import type { Request, Response, NextFunction } from 'express';
import { SECRET_KEY, ACCESS_TOKEN_TTL } from './config.js';
import { db, type Row } from './db.js';

const PBKDF2_ITERATIONS = 200_000;

/** Định dạng: pbkdf2$<iters>$<salt_hex>$<digest_hex> — y hệt Python hashlib. */
export function hashPassword(password: string): string {
  const salt = crypto.randomBytes(16).toString('hex');
  const digest = crypto.pbkdf2Sync(password, Buffer.from(salt, 'hex'),
    PBKDF2_ITERATIONS, 32, 'sha256').toString('hex');
  return `pbkdf2$${PBKDF2_ITERATIONS}$${salt}$${digest}`;
}

export function verifyPassword(password: string, stored: string): boolean {
  try {
    const [, iters, salt, digest] = stored.split('$');
    const cand = crypto.pbkdf2Sync(password, Buffer.from(salt, 'hex'),
      parseInt(iters, 10), digest.length / 2, 'sha256').toString('hex');
    const a = Buffer.from(cand), b = Buffer.from(digest);
    return a.length === b.length && crypto.timingSafeEqual(a, b);
  } catch { return false; }
}

export interface AccessClaims { sub: string; role: string; tv: number; iat: number; exp: number; }
export interface PendingClaims { sub: string; typ: string; iat: number; exp: number; }

export function createToken(userId: string, role: string, tokenVersion = 0): string {
  return jwt.sign({ sub: userId, role, tv: tokenVersion }, SECRET_KEY,
    { algorithm: 'HS256', expiresIn: ACCESS_TOKEN_TTL });
}
export function createPending2faToken(userId: string): string {
  return jwt.sign({ sub: userId, typ: '2fa' }, SECRET_KEY, { algorithm: 'HS256', expiresIn: 300 });
}
export function decodeToken(token: string): any | null {
  try { return jwt.verify(token, SECRET_KEY, { algorithms: ['HS256'] }); }
  catch { return null; }
}

function extractToken(req: Request): string | null {
  const h = req.headers.authorization || '';
  return h.startsWith('Bearer ') ? h.slice(7).trim() : null;
}
export function loadUser(userId: string): Row | null {
  return (db.prepare('SELECT * FROM users WHERE id=?').get(userId) as Row) || null;
}
function tokenFresh(payload: any, user: Row): boolean {
  return Number(payload.tv || 0) === Number(user.token_version || 0);
}

/** Ném lỗi HTTP (bắt ở middleware lỗi tập trung). */
export class HttpError extends Error {
  status: number; detail: any;
  constructor(status: number, detail: any) { super(typeof detail === 'string' ? detail : 'error'); this.status = status; this.detail = detail; }
}

// Gắn user đã xác thực vào req
declare global { namespace Express { interface Request { user?: Row; } } }

export function currentUser(req: Request): Row {
  const token = extractToken(req);
  const payload = token ? decodeToken(token) : null;
  if (!payload || payload.typ) throw new HttpError(401, 'Chưa đăng nhập hoặc phiên đã hết hạn');
  const user = loadUser(payload.sub);
  if (!user) throw new HttpError(401, 'Tài khoản không tồn tại');
  if (!tokenFresh(payload, user)) throw new HttpError(401, 'Phiên đã hết hạn — đăng nhập lại');
  return user;
}
export function optionalUser(req: Request): Row | null {
  const token = extractToken(req);
  const payload = token ? decodeToken(token) : null;
  if (!payload || payload.typ) return null;
  const user = loadUser(payload.sub);
  if (user && !tokenFresh(payload, user)) return null;
  return user;
}

// ---- RBAC middleware ----
export const STAFF_ROLES = ['admin', 'manager', 'uploader'];
export const MANAGER_ROLES = ['admin', 'manager'];
export const isManager = (u: Row): boolean => MANAGER_ROLES.includes(u.role);

export const requireAuth = (req: Request, _res: Response, next: NextFunction) => {
  try { req.user = currentUser(req); next(); } catch (e) { next(e); }
};
export const requireStaff = (req: Request, _res: Response, next: NextFunction) => {
  try { const u = currentUser(req); if (!STAFF_ROLES.includes(u.role)) throw new HttpError(403, 'Cần quyền quản trị nội dung'); req.user = u; next(); } catch (e) { next(e); }
};
export const requireManager = (req: Request, _res: Response, next: NextFunction) => {
  try { const u = currentUser(req); if (!MANAGER_ROLES.includes(u.role)) throw new HttpError(403, 'Cần quyền quản lý phát hành'); req.user = u; next(); } catch (e) { next(e); }
};
export const requireAdmin = (req: Request, _res: Response, next: NextFunction) => {
  try { const u = currentUser(req); if (u.role !== 'admin') throw new HttpError(403, 'Cần quyền quản trị'); req.user = u; next(); } catch (e) { next(e); }
};
export const attachOptionalUser = (req: Request, _res: Response, next: NextFunction) => {
  req.user = optionalUser(req) || undefined; next();
};
